transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -365,174 +307,124 @@ private void invokeExceptionCaught(final Throwable cause) {
 
     @Override
     public ChannelHandlerContext fireUserEventTriggered(final Object event) {
-        invokeUserEventTriggered(findContextInbound(MASK_USER_EVENT_TRIGGERED), event);
-        return this;
-    }
-
-    static void invokeUserEventTriggered(final AbstractChannelHandlerContext next, final Object event) {
         ObjectUtil.checkNotNull(event, "event");
-        EventExecutor executor = next.executor();
-        if (executor.inEventLoop()) {
-            next.invokeUserEventTriggered(event);
-        } else {
-            executor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    next.invokeUserEventTriggered(event);
-                }
-            });
-        }
-    }
-
-    private void invokeUserEventTriggered(Object event) {
-        if (invokeHandler()) {
-            try {
-                // DON'T CHANGE
-                // Duplex handlers implements both out/in interfaces causing a scalability issue
-                // see https://bugs.openjdk.org/browse/JDK-8180450
-                final ChannelHandler handler = handler();
-                final DefaultChannelPipeline.HeadContext headContext = pipeline.head;
-                if (handler == headContext) {
-                    headContext.userEventTriggered(this, event);
-                } else if (handler instanceof ChannelInboundHandlerAdapter) {
-                    ((ChannelInboundHandlerAdapter) handler).userEventTriggered(this, event);
-                } else {
-                    ((ChannelInboundHandler) handler).userEventTriggered(this, event);
+        AbstractChannelHandlerContext next = findContextInbound(MASK_USER_EVENT_TRIGGERED);
+        if (next.executor().inEventLoop()) {
+            if (next.invokeHandler()) {
+                try {
+                    // DON'T CHANGE
+                    // Duplex handlers implements both out/in interfaces causing a scalability issue
+                    // see https://bugs.openjdk.org/browse/JDK-8180450
+                    final ChannelHandler handler = next.handler();
+                    final DefaultChannelPipeline.HeadContext headContext = next.pipeline.head;
```

Comment:
this is still not great - here we have to chase 3 hops i.e. data dependent loads *just* to verify which is the `headContext`